contracts/nft/test/UniversalNFTCoreTest.sol

```diff
@@ -0,0 +1,167 @@
+import "../contracts/zetachain/UniversalNFT.sol";
+import "../contracts/shared/UniversalNFTEvents.sol";
+
+import "@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol";
+import "@uniswap/v2-periphery/contracts/interfaces/IUniswapV2Router02.sol";
+
+contract UniversalNFTIntegrationTest is
+    Test,
+    IGatewayZEVMEvents,
+    UniversalNFTEvents
+{
+    address payable proxy;
+    GatewayZEVM gateway;
+    ZRC20 zrc20;
+    WETH9 zetaToken;
+    SystemContract systemContract;
+    address owner;
+    address receiver;
+    address protocolAddress;
+    RevertOptions revertOptions;
+    CallOptions callOptions;
+    IUniswapV2Factory factory;
+    IUniswapV2Router02 router;
+
+    function setUp() public {
+        owner = address(this);
+        receiver = address(0x1234);
+        zetaToken = new WETH9();
+
+        vm.deal(owner, 100 ether);
+
+        factory = IUniswapV2Factory(
+            deployCode(
+                "out/UniswapV2Factory.sol/UniswapV2Factory.json",
+                abi.encode(address(0x1))
+            )
+        );
+
+        router = IUniswapV2Router02(
+            deployCode(
+                "out/UniswapV2Router02.sol/UniswapV2Router02.json",
+                abi.encode(address(factory), address(zetaToken))
+            )
+        );
+
+        proxy = payable(
+            Upgrades.deployUUPSProxy(
+                "GatewayZEVM.sol",
+                abi.encodeCall(
+                    GatewayZEVM.initialize,
+                    (address(zetaToken), owner)
+                )
+            )
+        );
+        gateway = GatewayZEVM(proxy);
+
+        protocolAddress = gateway.PROTOCOL_ADDRESS();
+
+        vm.startPrank(protocolAddress);
+        systemContract = new SystemContract(address(0), address(0), address(0));
+        zrc20 = new ZRC20(
+            "TOKEN",
+            "TKN",
+            18,
+            1,
+            CoinType.Gas,
+            0,
+            address(systemContract),
+            address(gateway)
+        );
+        systemContract.setGasCoinZRC20(1, address(zrc20));
+        systemContract.setGasPrice(1, 1);
+
+        vm.deal(protocolAddress, 100 ether);
+        zetaToken.deposit{value: 10 ether}();
+        zetaToken.approve(address(gateway), 10 ether);
+
+        zrc20.deposit(owner, 100 ether);
+
+        vm.stopPrank();
+
+        vm.startPrank(owner);
+
+        zrc20.approve(address(gateway), 100 ether);
+
+        zetaToken.deposit{value: 10 ether}();
+        zetaToken.approve(address(gateway), 10 ether);
+
+        zetaToken.approve(address(router), 10 ether);
+        zrc20.approve(address(router), 10 ether);
+
+        factory.createPair(address(zetaToken), address(zrc20));
+
+        router.addLiquidity( // throws error
+            address(zetaToken),
+            address(zrc20),
+            10 ether,
+            10 ether,
+            1 ether,
+            1 ether,
+            owner,
+            block.timestamp + 300
+        );
```

Comment:
@skosito I'm getting errors when trying to add liquidity. Is there anything wrong I'm doing here?